httpbp/response.go

```diff
@@ -1,313 +1,173 @@
+//
+// Response is responsible for setting the values that are independant of the
+// body as well as the ContentWriter to write the response body.
+type Response interface {
+	// SetCode sets the status code to return with the response.
+	SetCode(code int)
+
+	// StatusCode returns the current status code set on the response.
+	StatusCode() int
+
+	// ClearCookies clears all cookies currently set on the response.
+	ClearCookies()
+
+	// SetCookie adds the cookie to the list of cookies to set on the response.
+	SetCookie(cookie *http.Cookie)
+
+	// Cookies returns the list of cookies current set on the response.
+	Cookies() []*http.Cookie
+
+	// Headers returns the list of headers to return to the client.
+	Headers() http.Header
+
+	// ContentWriter returns the ContentWriter used to write the response.
+	ContentWriter() ContentWriter
 }
 
-// ResponseMessage returns the error message to send to the client.
-func (e HTTPError) ResponseMessage() string {
-	if e.Message != "" {
-		return e.Message
+// JSONContentWriter is a ContentWriterFactory that returns a ContentWriter
+// for writing JSON.
+//
+// When using a JSON content writer, your handler should return a value that
+// can be marshalled into JSON.  This can either be a struct that defines JSON
+// reflection tags or a `map` of values that can be Marshalled to JSON.
+func JSONContentWriter() ContentWriter {
+	return contentWriter{
+		contentType: JSONContentType,
+		write: func(w io.Writer, body interface{}) error {
+			return json.NewEncoder(w).Encode(body)
+		},
 	}
-	return http.StatusText(e.StatusCode())
 }
 
-// As implements helper interface for errors.As.
+// HTMLContentWriterFactory returns a ContentWriterFactory that returns a
+// ContentWriter for writing HTML using the given template.
 //
-// If v is pointer to either HTTPError or *HTTPError, *v will be set into this
-// error.
-func (e HTTPError) As(v interface{}) bool {
-	if target, ok := v.(*HTTPError); ok {
-		*target = e
-		return true
-	} else if target, ok := v.(**HTTPError); ok {
-		*target = &e
-		return true
+// When using an HTML content writer, your handler should return a struct that
+// can be given as input to t.Execute.
+func HTMLContentWriterFactory(t *template.Template) ContentWriterFactory {
+	return func() ContentWriter {
+		return contentWriter{
+			contentType: HTMLContentType,
+			write: func(w io.Writer, body interface{}) error {
+				return t.Execute(w, body)
+			},
+		}
 	}
-	return false
 }
 
-// Unwrap implements helper interface for errors.Unwrap.  Returns the optional
-// e.Cause error.
-func (e HTTPError) Unwrap() error {
-	return e.Cause
+// TextContentWriter is a ContentWriterFactory that returns a ContentWriter
+// for writing plain text.
+//
+// When using a text content writer, your handler should return a string,
+// WriteResponse will return an error rather than writing the content if you
+// pass a value of a different type.
+func TextContentWriter() ContentWriter {
```

Comment:
I think this should be `RawContentWriter` with the ability to specify content type instead, and the body should be `io.Reader` instead of `string`. Or maybe keep this one and add a `RawContentWriter`.